mtt/src/mtt.rs

```diff
@@ -0,0 +1,688 @@
+        };
+        let instance = Self {
+            inner: Mutex::new(MttL2Directory { mtt_l2_base }),
+        };
+        instance.validate_mtt(l2_addr_bits)?;
+        Ok(instance)
+    }
+
+    #[allow(clippy::drop_non_drop)]
+    fn is_range_of_type(
+        &self,
+        addr: SupervisorPageAddr,
+        len: usize,
+        memory_type: MttMemoryType,
+    ) -> Result<bool> {
+        let l2_dir = self.inner.lock();
+        let mut is_range_type_same = true;
+        let mut callback =
+            |addr: SupervisorPageAddr, _page_size: PageSize, l2_entry: MttL2Entry| {
+                let region_memory_type = MttL2Directory::get_memory_type(addr, &l2_entry);
+                if region_memory_type != memory_type {
+                    is_range_type_same = false;
+                    ControlFlow::Break(Ok(()))
+                } else {
+                    ControlFlow::Continue(true)
+                }
+            };
+
+        let mut mtt_range = MttRange::new(addr, len, l2_dir)?;
+        mtt_range.range_iterator(&mut callback)?;
+        drop(callback);
+        Ok(is_range_type_same)
+    }
+
+    fn page_size_from_entry(l2_entry: &MttL2Entry) -> PageSize {
+        use MttL2Entry::*;
+        match l2_entry {
+            NonConfidential1G(_) | Confidential1G(_) => PageSize::Size1G,
+            Mixed2M(_) => PageSize::Size2M,
+            MttL2Entry::L1Entry(_) => PageSize::Size4k,
+        }
+    }
+
+    #[allow(clippy::drop_non_drop)]
+    fn set_range_type(
+        &mut self,
+        addr: SupervisorPageAddr,
+        len: usize,
+        memory_type: MttMemoryType,
+        get_l1_page: &mut dyn FnMut() -> Option<Page<InternalClean>>,
+    ) -> Result<bool> {
+        let l2_dir = self.inner.lock();
+
+        let mut mtt_needs_invalidation = false;
+        let mut callback = |addr: SupervisorPageAddr, page_size: PageSize, l2_entry: MttL2Entry| {
+            let region_memory_type = MttL2Directory::get_memory_type(addr, &l2_entry);
+            if region_memory_type != memory_type {
+                mtt_needs_invalidation = true;
+                let target_page_size = Mtt::page_size_from_entry(&l2_entry);
+                if target_page_size == page_size {
+                    MttL2Directory::set_memory_type(addr, l2_entry, memory_type);
+                } else {
+                    // We are going from 1GB -> 64MB or 64MB -> L1, so
+                    // just split to the level below, and retry
+                    match l2_entry {
+                        MttL2Entry::NonConfidential1G(nc_1g) => nc_1g.split(),
+                        MttL2Entry::Confidential1G(conf_1g) => conf_1g.split(),
+                        MttL2Entry::Mixed2M(mixed_2m) => {
+                            if let Some(l1_page) = get_l1_page() {
+                                mixed_2m.split_to_l1_entry(l1_page)
+                            } else {
+                                return ControlFlow::Break(Err(InsufficientL1Pages));
```

Comment:
if we fail here, we might leave the range in a half-converted state. we probably need to do this in two passes: once to do the splitting, and once to make the conf/non-conf conversion.